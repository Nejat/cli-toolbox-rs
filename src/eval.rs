/// The `eval!` macro provides conditional code evaluation according to [`Verbosity`] level.
///
/// # Features
///
/// * you can evaluate code according to level of verbosity,
///   _i.e._ `Quite`, `Terse` _or_ `Verbose`
///
/// * all `Terse` evaluations will also evaluate if the level is set to `Verbose`
///
/// * you can choose unique evaluations for `Terse` or `Verbose` individually
///
/// # Basic Examples
///
/// * _evaluates if [`Verbosity`] is_ `Quite`
///
/// ```rust
/// # use cli_toolbox::eval;
/// eval! { QUITE foo() }
/// # fn foo() { /* ... */ }
/// ```
///
/// * _evaluates if [`Verbosity`] is_ `Terse` or `Verbose`
///
/// ```rust
/// # use cli_toolbox::eval;
/// eval! { TERSE foo() }
/// # fn foo() { /* ... */ }
/// ```
///
/// * _evaluates if [`Verbosity`] is_ `Verbose`
///
/// ```rust
/// # use cli_toolbox::eval;
/// eval! { VERBOSE foo() }
/// # fn foo() { /* ... */ }
/// ```
///
/// # Terse or Verbose Examples
///
/// These examples evaluate conditionally based on [`Verbosity`] level.
///
/// * _terse or verbose code blocks_
///
/// ```rust
/// # use cli_toolbox::eval;
/// eval! {
///     TERSE   { foo(); }
///     VERBOSE { junk(); }
/// }
/// # fn foo() { /* ... */ }
/// # fn junk() { /* ... */ }
/// ```
///
/// * _terse or verbose expressions_
///
/// ```rust
/// # use cli_toolbox::eval;
/// eval! {
///     TERSE   foo(),
///     VERBOSE junk()
/// }
/// # fn foo() { /* ... */ }
/// # fn junk() { /* ... */ }
/// ```
///
/// * _terse expression or verbose block_
///
/// ```rust
/// # use cli_toolbox::eval;
/// eval! {
///     TERSE   foo(),
///     VERBOSE { junk(); }
/// }
/// # fn foo() { /* ... */ }
/// # fn junk() { /* ... */ }
/// ```
///
/// * _terse block or verbose expression_
///
/// ```rust
/// # use cli_toolbox::eval;
/// eval! {
///     TERSE   { foo(); }
///     VERBOSE junk()
/// }
/// # fn foo() { /* ... */ }
/// # fn junk() { /* ... */ }
/// ```
///
/// [`Verbosity`]: crate::verbosity::Verbosity
#[macro_export]
macro_rules! eval {
    // ===============================================================================
    // evaluate if quite
    (QUITE $expr:expr) => {
        if $crate::Verbosity::is_quite() { $expr }
    };
    // ===============================================================================
    // evaluate if terse
    (TERSE $expr:expr) => {
        if $crate::Verbosity::is_terse() { $expr }
    };
    // ===============================================================================
    // evaluate if verbose
    (VERBOSE $expr:expr) => {
        if $crate::Verbosity::is_verbose() { $expr }
    };
    // ===============================================================================
    // evaluate terse or verbose blocks
    (
        TERSE   $terse_block:block
        VERBOSE $verbose_block:block
    ) => {
        match $crate::Verbosity::level() {
            $crate::Verbosity::Terse   => $terse_block
            $crate::Verbosity::Verbose => $verbose_block
            _ => {}
        }
    };
    // evaluate terse or verbose expressions
    (
        TERSE   $terse_expr:expr,
        VERBOSE $verbose_expr:expr
    ) => {
        match $crate::Verbosity::level() {
            $crate::Verbosity::Terse   => $terse_expr,
            $crate::Verbosity::Verbose => $verbose_expr,
            _ => {}
        }
    };
    // evaluate terse block or verbose expression
    (
        TERSE   $terse_block:block
        VERBOSE $verbose_expr:expr
    ) => {
        match $crate::Verbosity::level() {
            $crate::Verbosity::Terse   => $terse_block
            $crate::Verbosity::Verbose => $verbose_expr,
            _ => {}
        }
    };
    // evaluate terse expression or verbose block
    (
        TERSE   $terse_expr:expr,
        VERBOSE $verbose_block:block
    ) => {
        match $crate::Verbosity::level() {
            $crate::Verbosity::Terse   => $terse_expr,
            $crate::Verbosity::Verbose => $verbose_block
            _ => {}
        }
    };
}
