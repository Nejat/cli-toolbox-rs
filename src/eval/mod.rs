use syn::Expr;
use verbosity::Verbosity;

mod parse;
mod tokenize;

pub struct Eval {
    expr: Expr,
    verbosity: Verbosity
}