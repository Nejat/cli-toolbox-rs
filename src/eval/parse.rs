use syn::parse::{Parse, ParseStream};

use crate::common::parse::{parse_expression, parse_verbosity};
use crate::eval::Eval;

impl Parse for Eval {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        Ok(Self {
            verbosity: parse_verbosity(input)?,
            expr: parse_expression(input, "eval")?,
        })
    }
}