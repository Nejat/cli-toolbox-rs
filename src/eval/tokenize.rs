use proc_macro2::TokenStream;
use quote::ToTokens;

use crate::common::tokenize::tokenize_verbosity_checked_expr;
use crate::eval::Eval;

impl ToTokens for Eval {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        tokens.extend(tokenize_verbosity_checked_expr(self.verbosity, &self.expr));
    }
}
