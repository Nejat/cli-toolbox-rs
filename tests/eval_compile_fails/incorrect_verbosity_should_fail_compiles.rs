use cli_toolbox::eval;

fn main() {
    eval! { @err-terse println!("this should fail") }

    eval! { @err-verbose println!("this should fail") }

    eval! { @foo println!("this should fail") }
}