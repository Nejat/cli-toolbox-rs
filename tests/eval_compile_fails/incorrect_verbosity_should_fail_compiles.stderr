error: invalid verbosity designation, use @terse, @verbose or leave blank for default level
 --> tests/eval_compile_fails/incorrect_verbosity_should_fail_compiles.rs:4:13
  |
4 |     eval! { @err-terse println!("this should fail") }
  |             ^

error: invalid verbosity designation, use @terse, @verbose or leave blank for default level
 --> tests/eval_compile_fails/incorrect_verbosity_should_fail_compiles.rs:6:13
  |
6 |     eval! { @err-verbose println!("this should fail") }
  |             ^

error: invalid verbosity designation, use @terse, @verbose or leave blank for default level
 --> tests/eval_compile_fails/incorrect_verbosity_should_fail_compiles.rs:8:13
  |
8 |     eval! { @foo println!("this should fail") }
  |             ^
